libs/wingsdk/src/target-sim/function.inflight.ts

```diff
@@ -9,63 +13,176 @@ import {
+    }
   }
 
   public async init(): Promise<FunctionAttributes> {
     return {};
   }
 
   public async cleanup(): Promise<void> {
-    await this.sandbox.cleanup();
+    await this.createBundlePromise;
+    for (const worker of this.workers) {
+      await worker.cleanup();
+    }
   }
 
   public async save(): Promise<void> {}
 
   public async invoke(payload: string): Promise<string> {
     return this.context.withTrace({
       message: `Invoke (payload=${JSON.stringify(payload)}).`,
-      activity: () => this.sandbox.call("handler", payload),
+      activity: async () => {
+        await this.createBundlePromise; // ensure inflight code is bundled before doing any work
+        const worker = this.findAvailableWorker();
+        if (!worker) {
+          throw new Error(
+            "Too many requests, the function has reach its concurrency limit."
+          );
+        }
+        return worker.call("handler", payload);
+      },
     });
   }
 
   public async invokeAsync(payload: string): Promise<void> {
     await this.context.withTrace({
       message: `InvokeAsync (payload=${JSON.stringify(payload)}).`,
       activity: async () => {
+        await this.createBundlePromise; // ensure inflight code is bundled before doing any work
+        const worker = this.findAvailableWorker();
+        if (!worker) {
+          throw new Error(
+            "Too many requests, the function has reach its concurrency limit."
+          );
+        }
         process.nextTick(() => {
-          void this.sandbox.call("handler", payload);
+          // If the call fails, we log the error and continue since we've already
+          // handed control back to the caller.
+          void worker.call("handler", payload).catch((e) => {
+            this.context.addTrace({
+              data: {
+                message: `InvokeAsync (payload=${JSON.stringify(payload)}).`,
+                status: "failure",
+                error: e,
+              },
+              type: TraceType.RESOURCE,
+              sourcePath: this.context.resourcePath,
+              sourceType: FUNCTION_FQN,
+              timestamp: new Date().toISOString(),
+            });
+          });
         });
       },
     });
   }
+
+  // Used internally by cloud.Queue to apply backpressure
+  public async hasAvailableWorkers(): Promise<boolean> {
+    return (
+      this.workers.length < this.maxWorkers ||
+      this.workers.some((w) => w.isAvailable())
+    );
+  }
+
+  private findAvailableWorker(): Sandbox | undefined {
+    const worker = this.workers.find((w) => w.isAvailable());
+    if (worker) {
+      return worker;
+    }
+
+    if (this.workers.length < this.maxWorkers) {
+      const newWorker = this.initWorker();
+      this.workers.push(newWorker);
+      return newWorker;
+    }
+
+    return undefined;
+  }
+
+  private initWorker(): Sandbox {
+    if (!this.bundledFile) {
+      throw new Error("Bundle not created");
+    }
```

Comment:
What if we centralize the bundling await here?

```suggestion
    await this.createBundlePromise;
```

Reply:
I think inside `Sandbox`.